File: sdk/python/payraider/resources.py
"""Resource classes for each API group."""
from __future__ import annotations

from typing import Any, Callable, Coroutine, List, Optional
from urllib.parse import quote

from .http import HttpClient
from .types import (
    AlertRule,
    Anchor,
    ApiKey,
    AuthTokens,
    ConvertResult,
    Corridor,
    CostEstimateResponse,
    CreateApiKeyResponse,
    LiquidityPool,
    NetworkInfo,
    PaginatedResponse,
    PaymentPrediction,
    Price,
    PriceCacheStats,
    PricesBatchResult,
    Proposal,
    Transaction,
    VerifiedAsset,
    Webhook,
)


def _enc(value: str) -> str:
    return quote(value, safe="")


class AnchorsResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def list(self, *, page: Optional[int] = None, limit: Optional[int] = None) -> PaginatedResponse[Anchor]:
        return await self._http.request("GET", "/api/anchors", params={"page": page, "limit": limit})

    async def get(self, anchor_id: str) -> Anchor:
        return await self._http.request("GET", f"/api/anchors/{_enc(anchor_id)}")

    async def get_by_account(self, account: str) -> Anchor:
        return await self._http.request("GET", f"/api/anchors/account/{_enc(account)}")


class CorridorsResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def list(self, *, page: Optional[int] = None, limit: Optional[int] = None) -> PaginatedResponse[Corridor]:
        return await self._http.request("GET", "/api/corridors", params={"page": page, "limit": limit})

    async def get(self, source: str, destination: str) -> Any:
        return await self._http.request("GET", f"/api/corridors/{_enc(source)}/{_enc(destination)}")


class PricesResource:
    """
    The backend has no path-param or no-arg "list all" route - every method
    here takes an explicit asset (or list of assets) and matches
    backend/src/api/price_feed.rs exactly. See the equivalent
    sdk/typescript/src/resources.ts fix for the full rationale.
    """

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def get(self, asset: str) -> Price:
        return await self._http.request("GET", "/api/prices", params={"asset": asset})

    async def batch(self, assets: List[str]) -> PricesBatchResult:
        return await self._http.request("GET", "/api/prices/batch", params={"assets": ",".join(assets)})

    async def convert_to_usd(self, asset: str, amount: float) -> ConvertResult:
        """The backend only ever converts to USD - there is no arbitrary asset-to-asset conversion."""
        return await self._http.request("GET", "/api/prices/convert", params={"asset": asset, "amount": amount})

    async def cache_stats(self) -> PriceCacheStats:
        return await self._http.request("GET", "/api/prices/cache-stats")


class CostCalculatorResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def estimate(self, source_asset: str, destination_asset: str, amount: float) -> CostEstimateResponse:
        return await self._http.request(
            "POST", "/api/cost-calculator/estimate",
            json={"source_asset": source_asset, "destination_asset": destination_asset, "amount": amount},
        )

    async def routes(self, source_asset: str, destination_asset: str, amount: float) -> CostEstimateResponse:
        return await self._http.request(
            "POST", "/api/cost-calculator/routes",
            json={"source_asset": source_asset, "destination_asset": destination_asset, "amount": amount},
        )


class AlertsResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def list_rules(self, *, page: Optional[int] = None, limit: Optional[int] = None) -> PaginatedResponse[AlertRule]:
        return await self._http.request("GET", "/api/alerts/rules", params={"page": page, "limit": limit})

    async def create_rule(self, name: str, condition: str, threshold: float) -> AlertRule:
        return await self._http.request(
            "POST", "/api/alerts/rules",
            json={"name": name, "condition": condition, "threshold": threshold},
        )

    async def update_rule(self, rule_id: str, **kwargs: Any) -> AlertRule:
        return await self._http.request("PUT", f"/api/alerts/rules/{_enc(rule_id)}", json=kwargs)

    async def delete_rule(self, rule_id: str) -> None:
        await self._http.request("DELETE", f"/api/alerts/rules/{_enc(rule_id)}")

    async def list_history(self, *, page: Optional[int] = None, limit: Optional[int] = None) -> Any:
        return await self._http.request("GET", "/api/alerts/history", params={"page": page, "limit": limit})


class WebhooksResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def list(self) -> List[Webhook]:
        return await self._http.request("GET", "/api/webhooks")

    async def create(self, url: str, events: List[str]) -> Webhook:
        return await self._http.request("POST", "/api/webhooks", json={"url": url, "events": events})

    async def get(self, webhook_id: str) -> Webhook:
        return await self._http.request("GET", f"/api/webhooks/{_enc(webhook_id)}")

    async def delete(self, webhook_id: str) -> None:
        await self._http.request("DELETE", f"/api/webhooks/{_enc(webhook_id)}")

    async def test(self, webhook_id: str) -> Any:
        return await self._http.request("POST", f"/api/webhooks/{_enc(webhook_id)}/test")


class ApiKeysResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def list(self) -> List[ApiKey]:
        return await self._http.request("GET", "/api/api-keys")

    async def create(self, name: str) -> CreateApiKeyResponse:
        return await self._http.request("POST", "/api/api-keys", json={"name": name})

    async def get(self, key_id: str) -> ApiKey:
        return await self._http.request("GET", f"/api/api-keys/{_enc(key_id)}")

    async def rotate(self, key_id: str) -> CreateApiKeyResponse:
        return await self._http.request("POST", f"/api/api-keys/{_enc(key_id)}/rotate")

    async def revoke(self, key_id: str) -> None:
        await self._http.request("DELETE", f"/api/api-keys/{_enc(key_id)}")


class AuthResource:
    def __init__(self, http: HttpClient, set_token: Callable[[str], None]) -> None:
        self._http = http
        self._set_token = set_token

    async def login(self, email: str, password: str) -> AuthTokens:
        tokens = await self._http.request("POST", "/api/auth/login", json={"email": email, "password": password})
        self._set_token(tokens["access_token"])
        return tokens

    async def refresh(self, refresh_token: str) -> AuthTokens:
        tokens = await self._http.request("POST", "/api/auth/refresh", json={"refresh_token": refresh_token})
        self._set_token(tokens["access_token"])
        return tokens

    async def logout(self) -> None:
        await self._http.request("POST", "/api/auth/logout")


class LiquidityPoolsResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def list(self, *, page: Optional[int] = None, limit: Optional[int] = None) -> PaginatedResponse[LiquidityPool]:
        return await self._http.request("GET", "/api/liquidity-pools", params={"page": page, "limit": limit})

    async def get(self, pool_id: str) -> LiquidityPool:
        return await self._http.request("GET", f"/api/liquidity-pools/{_enc(pool_id)}")


class TransactionsResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def create(self, envelope_xdr: str) -> Transaction:
        return await self._http.request("POST", "/api/transactions", json={"envelope_xdr": envelope_xdr})

    async def get(self, tx_id: str) -> Transaction:
        return await self._http.request("GET", f"/api/transactions/{_enc(tx_id)}")

    async def submit(self, tx_id: str) -> Transaction:
        return await self._http.request("POST", f"/api/transactions/{_enc(tx_id)}/submit")


class NetworkResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def info(self) -> NetworkInfo:
        return await self._http.request("GET", "/api/network")

    async def available(self) -> List[NetworkInfo]:
        return await self._http.request("GET", "/api/network/available")


class MlResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def predict(self, **params: Any) -> PaymentPrediction:
        return await self._http.request("POST", "/api/ml/predict", json=params)

    async def model_status(self) -> Any:
        return await self._http.request("GET", "/api/ml/status")


class GovernanceResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def list_proposals(self, *, page: Optional[int] = None, limit: Optional[int] = None) -> PaginatedResponse[Proposal]:
        return await self._http.request("GET", "/api/governance/proposals", params={"page": page, "limit": limit})

    async def create_proposal(self, title: str, description: str) -> Proposal:
        return await self._http.request("POST", "/api/governance/proposals", json={"title": title, "description": description})

    async def get_proposal(self, proposal_id: str) -> Proposal:
        return await self._http.request("GET", f"/api/governance/proposals/{_enc(proposal_id)}")

    async def vote(self, proposal_id: str, support: bool) -> Any:
        return await self._http.request("POST", f"/api/governance/proposals/{_enc(proposal_id)}/vote", json={"support": support})


class AssetVerificationResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def verify(self, asset_code: str, asset_issuer: str) -> VerifiedAsset:
        return await self._http.request(
            "POST", "/api/asset-verification/verify",
            json={"asset_code": asset_code, "asset_issuer": asset_issuer},
        )

    async def get(self, asset_code: str, asset_issuer: str) -> VerifiedAsset:
        return await self._http.request("GET", f"/api/asset-verification/{_enc(asset_code)}/{_enc(asset_issuer)}")

    async def list(self, *, page: Optional[int] = None, limit: Optional[int] = None) -> PaginatedResponse[VerifiedAsset]:
        return await self._http.request("GET", "/api/asset-verification", params={"page": page, "limit": limit})

File: sdk/typescript/src/resources.ts
import type { HttpClient } from "./http.js";
import type {
  Anchor,
  AlertRule,
  ApiKey,
  AuthTokens,
  ConvertResult,
  CorridorDetail,
  Corridor,
  CostEstimateRequest,
  CostEstimateResponse,
  CreateAlertRuleRequest,
  CreateApiKeyRequest,
  CreateApiKeyResponse,
  CreateProposalRequest,
  CreateTransactionRequest,
  CreateWebhookRequest,
  LiquidityPool,
  LoginRequest,
  NetworkInfo,
  PaginatedResponse,
  PaginationParams,
  PaymentPrediction,
  Price,
  Proposal,
  Transaction,
  VerifiedAsset,
  Webhook,
  VoteTally,
  GovernanceProposal,
} from "./types.js";

export class AnchorsResource {
  constructor(private http: HttpClient) {}

  list(params?: PaginationParams): Promise<PaginatedResponse<Anchor>> {
    return this.http.request("GET", "/api/anchors", { params: params as Record<string, unknown> });
  }

  get(id: string): Promise<Anchor> {
    return this.http.request("GET", `/api/anchors/${encodeURIComponent(id)}`);
  }

  getByAccount(account: string): Promise<Anchor> {
    return this.http.request("GET", `/api/anchors/account/${encodeURIComponent(account)}`);
  }
}

export class CorridorsResource {
  constructor(private http: HttpClient) {}

  list(params?: PaginationParams): Promise<PaginatedResponse<Corridor>> {
    return this.http.request("GET", "/api/corridors", { params: params as Record<string, unknown> });
  }

  get(source: string, destination: string): Promise<CorridorDetail> {
    return this.http.request(
      "GET",
      `/api/corridors/${encodeURIComponent(source)}/${encodeURIComponent(destination)}`,
    );
  }
}

export class PricesResource {
  constructor(private http: HttpClient) {}

  list(): Promise<Price[]> {
    return this.http.request("GET", "/api/prices");
  }

  get(asset: string): Promise<Price> {
    return this.http.request("GET", `/api/prices/${encodeURIComponent(asset)}`);
  }

  convert(from: string, to: string, amount: number): Promise<ConvertResult> {
    return this.http.request("POST", "/api/prices/convert", {
      body: { from_asset: from, to_asset: to, amount },
    });
  }
}

export class CostCalculatorResource {
  constructor(private http: HttpClient) {}

  estimate(req: CostEstimateRequest): Promise<CostEstimateResponse> {
    return this.http.request("POST", "/api/cost-calculator/estimate", { body: req });
  }

  routes(req: CostEstimateRequest): Promise<CostEstimateResponse> {
    return this.http.request("POST", "/api/cost-calculator/routes", { body: req });
  }
}

export class AlertsResource {
  constructor(private http: HttpClient) {}

  listRules(params?: PaginationParams): Promise<PaginatedResponse<AlertRule>> {
    return this.http.request("GET", "/api/alerts/rules", { params: params as Record<string, unknown> });
  }

  createRule(req: CreateAlertRuleRequest): Promise<AlertRule> {
    return this.http.request("POST", "/api/alerts/rules", { body: req });
  }

  updateRule(id: string, req: Partial<CreateAlertRuleRequest>): Promise<AlertRule> {
    return this.http.request("PUT", `/api/alerts/rules/${encodeURIComponent(id)}`, { body: req });
  }

  deleteRule(id: string): Promise<void> {
    return this.http.request("DELETE", `/api/alerts/rules/${encodeURIComponent(id)}`);
  }

  listHistory(params?: PaginationParams): Promise<PaginatedResponse<unknown>> {
    return this.http.request("GET", "/api/alerts/history", { params: params as Record<string, unknown> });
  }
}

export class WebhooksResource {
  constructor(private http: HttpClient) {}

  list(): Promise<Webhook[]> {
    return this.http.request("GET", "/api/webhooks");
  }

  create(req: CreateWebhookRequest): Promise<Webhook> {
    return this.http.request("POST", "/api/webhooks", { body: req });
  }

  get(id: string): Promise<Webhook> {
    return this.http.request("GET", `/api/webhooks/${encodeURIComponent(id)}`);
  }

  delete(id: string): Promise<void> {
    return this.http.request("DELETE", `/api/webhooks/${encodeURIComponent(id)}`);
  }

  test(id: string): Promise<unknown> {
    return this.http.request("POST", `/api/webhooks/${encodeURIComponent(id)}/test`);
  }
}

export class ApiKeysResource {
  constructor(private http: HttpClient) {}

  list(): Promise<ApiKey[]> {
    return this.http.request("GET", "/api/api-keys");
  }

  create(req: CreateApiKeyRequest): Promise<CreateApiKeyResponse> {
    return this.http.request("POST", "/api/api-keys", { body: req });
  }

  get(id: string): Promise<ApiKey> {
    return this.http.request("GET", `/api/api-keys/${encodeURIComponent(id)}`);
  }

  rotate(id: string): Promise<CreateApiKeyResponse> {
    return this.http.request("POST", `/api/api-keys/${encodeURIComponent(id)}/rotate`);
  }

  revoke(id: string): Promise<void> {
    return this.http.request("DELETE", `/api/api-keys/${encodeURIComponent(id)}`);
  }
}

export class AuthResource {
  constructor(
    private http: HttpClient,
    private setToken: (token: string) => void,
  ) {}

  async login(req: LoginRequest): Promise<AuthTokens> {
    const tokens = await this.http.request<AuthTokens>("POST", "/api/auth/login", { body: req });
    this.setToken(tokens.access_token);
    return tokens;
  }

  async refresh(refreshToken: string): Promise<AuthTokens> {
    const tokens = await this.http.request<AuthTokens>("POST", "/api/auth/refresh", {
      body: { refresh_token: refreshToken },
    });
    this.setToken(tokens.access_token);
    return tokens;
  }

  logout(): Promise<void> {
    return this.http.request("POST", "/api/auth/logout");
  }
}

export class LiquidityPoolsResource {
  constructor(private http: HttpClient) {}

  list(params?: PaginationParams): Promise<PaginatedResponse<LiquidityPool>> {
    return this.http.request("GET", "/api/liquidity-pools", { params: params as Record<string, unknown> });
  }

  get(id: string): Promise<LiquidityPool> {
    return this.http.request("GET", `/api/liquidity-pools/${encodeURIComponent(id)}`);
  }
}

export class TransactionsResource {
  constructor(private http: HttpClient) {}

  create(req: CreateTransactionRequest): Promise<Transaction> {
    return this.http.request("POST", "/api/transactions", { body: req });
  }

  get(id: string): Promise<Transaction> {
    return this.http.request("GET", `/api/transactions/${encodeURIComponent(id)}`);
  }

  submit(id: string): Promise<Transaction> {
    return this.http.request("POST", `/api/transactions/${encodeURIComponent(id)}/submit`);
  }
}

export class NetworkResource {
  constructor(private http: HttpClient) {}

  info(): Promise<NetworkInfo> {
    return this.http.request("GET", "/api/network");
  }

  available(): Promise<NetworkInfo[]> {
    return this.http.request("GET", "/api/network/available");
  }
}

export class MlResource {
  constructor(private http: HttpClient) {}

  predict(params: Record<string, unknown>): Promise<PaymentPrediction> {
    return this.http.request("POST", "/api/ml/predict", { body: params });
  }

  modelStatus(): Promise<Record<string, unknown>> {
    return this.http.request("GET", "/api/ml/status");
  }
}

export class GovernanceResource {
  constructor(private http: HttpClient) {}

  listProposals(params?: PaginationParams): Promise<PaginatedResponse<Proposal>> {
    return this.http.request("GET", "/api/governance/proposals", { params: params as Record<string, unknown> });
  }

  createProposal(req: CreateProposalRequest): Promise<Proposal> {
    return this.http.request("POST", "/api/governance/proposals", { body: req });
  }

  getProposal(id: string): Promise<Proposal> {
    return this.http.request("GET", `/api/governance/proposals/${encodeURIComponent(id)}`);
  }

  vote(id: string, support: boolean): Promise<VoteTally> {
    return this.http.request("POST", `/api/governance/proposals/${encodeURIComponent(id)}/vote`, {
      body: { support },
    });
  }
}

export class AssetVerificationResource {
  constructor(private http: HttpClient) {}

  verify(assetCode: string, assetIssuer: string): Promise<VerifiedAsset> {
    return this.http.request("POST", "/api/asset-verification/verify", {
      body: { asset_code: assetCode, asset_issuer: assetIssuer },
    });
  }

  get(assetCode: string, assetIssuer: string): Promise<VerifiedAsset> {
    return this.http.request(
      "GET",
      `/api/asset-verification/${encodeURIComponent(assetCode)}/${encodeURIComponent(assetIssuer)}`,
    );
  }

  list(params?: PaginationParams): Promise<PaginatedResponse<VerifiedAsset>> {
    return this.http.request("GET", "/api/asset-verification", { params: params as Record<string, unknown> });
  }
}
